tensorboard/plugins/image/polymer3/tf_image_dashboard/tf-image-loader.ts

```diff
@@ -0,0 +1,364 @@
+    return steps.length - 1;
+  }
+  @computed('sample')
+  get _sampleText(): string {
+    var sample = this.sample;
+    return `${sample + 1}`;
+  }
+  @computed('ofSamples')
+  get _hasMultipleSamples(): boolean {
+    var ofSamples = this.ofSamples;
+    return ofSamples > 1;
+  }
+  _getAriaExpanded() {
+    return this.actualSize ? 'true' : 'false';
+  }
+  attached() {
+    this.reload();
+  }
+  @observe('run', 'tag')
+  reload() {
+    if (!this.isAttached) {
+      return;
+    }
+    this._metadataCanceller.cancelAll();
+    const router = getRouter();
+    const url = addParams(router.pluginRoute('images', '/images'), {
+      tag: this.tag,
+      run: this.run,
+      sample: this.sample as any,
+    });
+    const updateSteps = this._metadataCanceller.cancellable((result) => {
+      if (result.cancelled) {
+        return;
+      }
+      const data = result.value as any;
+      const steps = data.map(this._createStepDatum.bind(this));
+      this.set('_steps', steps);
+      this.set('_stepIndex', steps.length - 1);
+    });
+    this.requestManager.request(url).then(updateSteps);
+  }
+  _createStepDatum(imageMetadata) {
+    let url = getRouter().pluginRoute('images', '/individualImage');
+    // Include wall_time just to disambiguate the URL and force
+    // the browser to reload the image when the URL changes. The
+    // backend doesn't care about the value.
+    url = addParams(url, {ts: imageMetadata.wall_time});
+    url += '&' + imageMetadata.query;
+    return {
+      // The wall time within the metadata is in seconds. The Date
+      // constructor accepts a time in milliseconds, so we multiply by 1000.
+      wall_time: new Date(imageMetadata.wall_time * 1000),
+      step: imageMetadata.step,
+      url,
+    };
+  }
+  @observe('_currentStep', 'brightnessAdjustment', 'contrastPercentage')
+  _updateImageUrl() {
+    var currentStep = this._currentStep;
+    var brightnessAdjustment = this.brightnessAdjustment;
+    var contrastPercentage = this.contrastPercentage;
+    // We manually change the image URL (instead of binding to the
+    // image's src attribute) because we would like to manage what
+    // happens when the image starts and stops loading.
+    if (!currentStep) return;
+    const img = new Image();
+    this._imageCanceller.cancelAll();
+    img.onload = img.onerror = this._imageCanceller
+      .cancellable((result) => {
+        if (result.cancelled) {
+          return;
+        }
+        const mainImageContainer = this.$$('#main-image-container');
+        mainImageContainer.textContent = '';
+        (PolymerDom.dom(mainImageContainer) as any).appendChild(img);
+        this.set('_isImageLoading', false);
+      })
+      .bind(this);
+    img.style.filter = `contrast(${contrastPercentage}%) `;
+    img.style.filter += `brightness(${brightnessAdjustment})`;
+    // Load the new image.
+    this.set('_isImageLoading', true);
+    img.src = (currentStep as any).url;
```

Comment:
Is tsetse okay with `img.src`?

Reply:
I guess it is—the sync (<http://cl/325324290>) builds and works. (That
is a bit interesting, though.)